staticd/static_vty.c

```diff
@@ -367,169 +158,156 @@ static int static_route_leak(
-						gate_str);
+		if (label_str) {
+			/* copy of label string (start) */
+			char *ostr;
+			/* pointer to next segment */
+			char *nump;
+
+			snprintf(xpath, sizeof(xpath),
+				 FRR_STATIC_ROUTE_NH_LABEL_XPATH, buf_nh_type,
+				 nh_svrf, buf_gate_str, ifname);
+			strlcpy(ab_xpath, xpath_list, sizeof(ab_xpath));
+			strlcat(ab_xpath, xpath, sizeof(ab_xpath));
+			nb_cli_enqueue_change(vty, ab_xpath, NB_OP_DESTROY,
+					      NULL);
+
+			ostr = XSTRDUP(MTYPE_TMP, label_str);
+			while ((nump = strsep(&ostr, "/")) != NULL) {
+				snprintf(xpath, sizeof(xpath),
+					 FRR_STATIC_ROUTE_NHLB_KEY_XPATH,
+					 buf_nh_type, nh_svrf, buf_gate_str,
+					 ifname, label_stack_id);
+				strlcpy(ab_xpath, xpath_list, sizeof(ab_xpath));
+				strlcat(ab_xpath, xpath, sizeof(ab_xpath));
+
+				nb_cli_enqueue_change(vty, ab_xpath,
+						      NB_OP_MODIFY, nump);
+				label_stack_id++;
+			}
+			XFREE(MTYPE_TMP, ostr);
+		} else {
+			snprintf(xpath, sizeof(xpath),
+				 FRR_STATIC_ROUTE_NH_LABEL_XPATH, buf_nh_type,
+				 nh_svrf, buf_gate_str, ifname);
+			strlcpy(ab_xpath, xpath_list, sizeof(ab_xpath));
+			strlcat(ab_xpath, xpath, sizeof(ab_xpath));
+			nb_cli_enqueue_change(vty, ab_xpath, NB_OP_DESTROY,
+					      NULL);
 		}
-
-	}
-
-	if (gate_str == NULL && ifname == NULL)
-		type = STATIC_BLACKHOLE;
-	else if (gate_str && ifname) {
-		if (afi == AFI_IP)
-			type = STATIC_IPV4_GATEWAY_IFNAME;
-		else
-			type = STATIC_IPV6_GATEWAY_IFNAME;
-	} else if (ifname)
-		type = STATIC_IFNAME;
-	else {
-		if (afi == AFI_IP)
-			type = STATIC_IPV4_GATEWAY;
-		else
-			type = STATIC_IPV6_GATEWAY;
-	}
-
-	if (!negate) {
-		static_add_route(afi, safi, type, &p, src_p, gatep, ifname,
-				 bh_type, tag, distance, svrf, nh_svrf,
-				 &snh_label, table_id, onlink);
-		/* Mark as having FRR configuration */
-		vrf_set_user_cfged(svrf->vrf);
+		ret = nb_cli_apply_changes(vty, xpath_list);
 	} else {
-		static_delete_route(afi, safi, type, &p, src_p, gatep, ifname,
-				    tag, distance, svrf, &snh_label, table_id);
-		/* If no other FRR config for this VRF, mark accordingly. */
-		if (!static_vrf_has_config(svrf))
-			vrf_reset_user_cfged(svrf->vrf);
+		if (src_str)
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_DEL_S_ROUTE_SRC_NH_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix,
+				 yang_afi_safi_value2identity(afi, safi),
+				 buf_src_prefix, distance, tag,
+				 table_id, buf_nh_type, nh_svrf, buf_gate_str,
+				 ifname);
+		else
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_DEL_S_ROUTE_NH_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix,
+				 yang_afi_safi_value2identity(afi, safi),
+				 distance, tag, table_id,
+				 buf_nh_type, nh_svrf, buf_gate_str, ifname);
+		dnode = yang_dnode_get(vty->candidate_config->dnode,
+				       xpath_list);
+		if (!dnode)
+			return ret;
+
+		dnode = yang_get_subtree_with_no_sibling(dnode);
```

Comment:
I understand the problem here but I got concerned with the complexity of `yang_get_subtree_with_no_sibling()`. Let's discuss this tomorrow and see if we can come up with a simpler solution.

Reply:
fine